# Built-in data types

In order to store information during the run of a program we use variables. Every variable has its type which determines the size and organization of the memory allocated to it. Java supports two kinds of data types - built-in types and objective ones. Built-in data types contain a number of types for representing integers (byte, short, int, long), two types for floating-point numbers (float and double), a type for representing characters (char) and a type for representing logical value (boolean). In order to store information during a run of a program, we need to create and use variables of these data types - this essentially means three actions:


- declaring (a type of) variable (1),
- assigning a value to it (2) and, finally,
- performing operations on it (3).

In order to declare a variable, you need to specify its type followed by the name which will be used to identify this variable.

In [19]:
int i;   //(1)

Assigning a value to a built-in variable is performed with the use of Java assignment operator "=".

In [ ]:
1=7;   //(2)

Finally, performing operations on a variable requires the use of operators.

In [18]:
i=i+3;   //(3)

## Numeric types - integers

Java uses four types for integers that differ with respect to the size of the storage allocated for the given type.

Data type|Range|Size
-|-|-
byte|$-2^7$ to $2^7-1$|8-bit signed
short|$-2^{15}$ to $2^{15}-1$|16-bit signed
int|$-2^{31}$ to $2^{31}-1$|32-bit signed
long|$-2^{63}$ to $2^{63}-1$|64-bit signed

A variable of type `int` can be assigned integer in the range from -2 147 483 648 to 2 147 483 647.

In [62]:
byte i = 127;
System.out.println(i);

127


In [63]:
short i = 32767;
System.out.println(i);

32767


In [65]:
int i = 2147483647;
System.out.println(i);

2147483647


An integer literal is assumed to be of the `int` type - to denote an integer literal of the `long` type it must be followed by "L" (or "l") sign.

In [69]:
long i = 9223372036854775807L;
System.out.println(i);

9223372036854775807


To improve readability, you may use underscores between two digits in a number literal.

In [70]:
long i = 9_223_372_036_854_775_807L;
System.out.println(i);

9223372036854775807


By default, an integer literal denotes a decimal integer, a binary integer literal it must be denoted with leading "0b" or "0B" (zero B).

In [75]:
int i = 0B1111;
System.out.println(i);

15


An octal integer literal is denoted by a leading "0" (zero).

In [76]:
int i = 07777;
System.out.println(i);

4095


 A hexadecimal integer literal is denoted with leading "0x" or "0X" (zero X).

In [77]:
int i = 0XFFFF;
System.out.println(i);

65535


In [85]:
int i;   //(1)
i=7;     //(2)
i=i+3;   //(3)

System.out.println(i);

10


## Numeric types - floating-point numbers

To represent floating point numbers, Java support two built-in data types complying with IEEE 754 (standard approved by the Institute of Electrical and Electronics Engineers for representing floating-point numbers).

Data type|Range|Precision|Size
-|-|-|-
float|approx. $\pm 3.4 * {10}^{38}$|7-8 significant digits|32-bit
double|approx. $\pm 1.8 * {10}^{308}$|15-17 significant digits|64-bit

A floating point literal is assumed to be of the `double` type - to denote a floating point literal of the `float` type it must be followed by "F" (or "f") sign.

In [74]:
float x = 3.14F;
System.out.println(x);

3.14


In [73]:
double x = 3.14;
System.out.println(x);

3.14


In [80]:
double x = 1.7976931348623157E+308;
System.out.println(x);

1.7976931348623157E308


In [86]:
double x;   //(1)
x=3.14;     //(2)
x=x*2;      //(3)

System.out.println(x);

6.28


A ```float``` is represented using 32 bits, and each possible combination of bits represents one real number. This means that at most $2^{32}$ possible real numbers can be exactly represented (including including plus infinity, minus infinity, and NaN), even though there are infinitely many real numbers, and is represented as 

$$(-1)^s \cdot M \cdot 2^{e-127}$$

with a sign bit s (interpreted as plus or minus), 8 bits for the exponent e, and 23 bits for the mantissa M. A ```double``` uses 64 bits - an 11 bit exponent with a bias of 1023, and a 52 bit mantissa.

Contrary to integer operations, floating point arithmetic is always approximate and the precision of the result depends on the data types used. As a result, floating point types must not be used to represent, for example, monetary data. Cumulation of floating point round off errors was the cause of a number of spectacular incidents like the explosion of unmanned Ariane 5 space rocket in 1996. This must be taken into account particularly when comparing floating point variables.

In [124]:
double x = 3*0.1;
double y = 0.3;

System.out.println(x == y);

false


In [125]:
double x = 5*0.1;
double y = 0.5;

System.out.println(x == y);

true


Or when it cames to floating point operations.

In [92]:
float x = 1.0F;
float y = 3.0F;
float z = x / y;

System.out.println(z);

0.33333334


In [94]:
double x = 1.0;
double y = 3.0;
double z = x / y;

System.out.println(z);

0.3333333333333333


Particular attention must be paid in the case of operations between the numbers which significantly differ in magnitude.

In [118]:
float x = 3.72467777E-3F;
float y = 7.37243333E+5F;

float z = x + y;

System.out.println(z);

737243.3


In [119]:
double x = 3.72467777E-3;
double y = 7.37243333E+5;

double z = x + y;

System.out.println(z);

737243.3367246777


## Character type

Java supports a two-byte built-in type that represents a single 16 bit Unicode (UTF-16) character

Data type|Range|Size
-|-|-
char|'\u0000' (0) to '\uffff' (65535)|16-bit

You can use such characters directly in your code or you can use a "Unicode escape" such as __```Se\u00F1or```__ (Señor in Spanish). Java also supports a number of special escape sequences: __```\b```__ (backspace), __```\t```__ (tab), __```\n```__ (line feed), __```\f```__ (form feed), __```\r```__ (carriage return), __```\"```__ (double quote), __```\'```__ (single quote), and __```\\```__ (backslash).

In [26]:
char c = 'A';
System.out.println(c);

A


In [33]:
char c='\u0042';
System.out.println(c);

B


An object of the Java API String class contains an array of the `char` type signs.

In [83]:
String s = "Text including sp\u0041ces";
System.out.println(s);

Text including spAces


In [87]:
char c;   //(1)
c='A';    //(2)
c++;      //(3)
System.out.println(c);

c+=3;     //(3)
System.out.println(c);

B
E


## Boolean type

Boolean values are represented by Java the built-in type `boolean`, which may have two values, denoted as `true` and `false`.

In [89]:
boolean b = true;
System.out.println(b);

true


In [90]:
boolean b = false;
System.out.println(b);

false


In [13]:
boolean b;   //(1)
b=false;     //(2)
b=!b;        //(3)

System.out.println(b);

true
